src/lib/update/clients/inst_update_partition_auto.rb

```diff
@@ -0,0 +1,154 @@
+
+      Yast.include self, "update/rootpart.rb"
+
+      # In case of restarting after a installer update, we restore previous
+      # data if exists (bsc#988287)
+      load_data if Installation.restarting? && data_stored?
+
+      if RootPart.Mounted
+        log.debug("RootPart is mounted, detaching Update & unmounting partitions")
+        Update.Detach
+        RootPart.UnmountPartitions(false)
+      end
+
+      RootPart.Detect
+
+      # if there is only one suitable partition which can be mounted, use it without asking
+      @target_system = current_target_system
+
+      if @target_system
+        log.info("Auto-mounting system located at #{@target_system}")
+
+        RootPart.selectedRootPartition = @target_system
+        RootPart.targetOk = RootPart.mount_target
+
+        # Not mounted correctly
+        if !RootPart.targetOk
+          # error report
+          Report.Error(_("Failed to mount target system"))
+          UmountMountedPartition()
+          # Correctly mounted but incomplete installation found
+        elsif RootPart.IncompleteInstallationDetected(Installation.destdir)
+          Report.Error(
+            _("A possibly incomplete installation has been detected.")
+          )
+          UmountMountedPartition()
+        elsif !(Pkg.TargetInitializeOptions(Installation.destdir,
+          "target_distro" => target_distribution) && Pkg.TargetLoad)
+
+          Report.Error("Initializing the target system failed")
+          UmountMountedPartition()
+          Pkg.TargetFinish
+        else
+          store_data
+
+          return :next
+        end
+      end
+
+      @ret = RootPartitionDialog(:update_dialog)
+
+      store_data if @ret == :next
+
+      @ret
+    end
+
+  private
+
+    # @return <Boolean> true if dumped file data exists.
+    def data_stored?
+      ::File.exist?(DATA_PATH)
+    end
+
+    # Save some important RootPart attributes into a yaml file.
+    def store_data
+      data = {
+        "activated"  => RootPart.GetActivated,
+        "selected"   => RootPart.selectedRootPartition,
+        "previous"   => RootPart.previousRootPartition,
+        "partitions" => RootPart.rootPartitions
+      }
+
+      File.write(DATA_PATH, data.to_yaml)
+    end
+
+    # Loads RootPart data from a dump yaml file and delete the file after that.
+    # It also remember the current root selection as the target_system
+    def load_data
+      data = YAML.load(File.read(DATA_PATH))
+
+      log.debug("Loading data from dump file: #{data}")
+      RootPart.load_saved(data)
+
+      root_target = RootPart.selectedRootPartition || ""
+
+      @target_system = root_target unless root_target.empty?
+
+      ::FileUtils.rm_rf(DATA_PATH)
+    end
+
+    # Obtains the target system from the install.inf file or use the current
+    # partitions if there is only 1 valid.
+    #
+    # @return [String] target root or nil
+    def current_target_system
+      return @target_system if Installation.restarting?
```

Comment:
In case of Restarting, it also remember the partition you choose before the self update. But i can move this knowledge/responsability out of the method.